core/src/main/scala/kafka/log/Log.scala

```diff
@@ -713,9 +713,15 @@ class Log(@volatile var dir: File,
           }
         } else {
           // we are taking the offsets we are given
-          if (!appendInfo.offsetsMonotonic || appendInfo.firstOrLastOffset < nextOffsetMetadata.messageOffset)
+          if (!appendInfo.offsetsMonotonic)
             throw new IllegalArgumentException(s"Out of order offsets found in append to $topicPartition: " +
               records.records.asScala.map(_.offset))
+          if (appendInfo.firstOrLastOffset < nextOffsetMetadata.messageOffset) {
+            val firstOrLast = if (appendInfo.firstOffset.isDefined) "First" else "Last"
+            throw new IllegalArgumentException(s"Out of order offsets found in append to $topicPartition. $firstOrLast " +
```

Comment:
I think the first sentence is still misleading since there are no out of order offsets in the append itself. Maybe "Attempt to append batch at invalid offset" or something like that?